// @blocknote/react/dist/FloatingComposerController-D-j-wzDn.js
import { t as e } from "./rolldown-runtime-CAFD8bLK.js";
import { f as t, i as n, l as r, m as i, n as a, o, p as s, r as c, s as l, t as u } from "./defaultCommentEditorSchema-BWkUaI17.js";
import { useMemo as d } from "react";
import { CommentsExtension as f } from "@blocknote/core/comments";
import { flip as p, offset as m, shift as h } from "@floating-ui/react";
import { mergeCSSClasses as g } from "@blocknote/core";
import { jsx as _ } from "react/jsx-runtime";
import { TextSelection as v } from "@tiptap/pm/state";
//#region src/components/Comments/FloatingComposer.tsx
function y() {
	let e = i(), n = t(f), o = r(), s = l(), d = c({
		trailingBlock: !1,
		dictionary: {
			...s,
			placeholders: { emptyDocument: s.placeholders.new_comment }
		},
		schema: n.commentEditorSchema || u
	});
	return /* @__PURE__ */ _(o.Comments.Card, {
		className: "bn-thread",
		children: /* @__PURE__ */ _(a, {
			autoFocus: !0,
			editable: !0,
			editor: d,
			actions: ({ isEmpty: t }) => /* @__PURE__ */ _(o.Generic.Toolbar.Root, {
				className: g("bn-action-toolbar", "bn-comment-actions"),
				variant: "action-toolbar",
				children: /* @__PURE__ */ _(o.Generic.Toolbar.Button, {
					className: "bn-button",
					mainTooltip: s.comments.save_button_text,
					variant: "compact",
					isDisabled: t,
					onClick: async () => {
						await n.createThread({ initialComment: { body: d.document } }), n.stopPendingComment(), e.transact((e) => {
							e.setSelection(v.create(e.doc, e.selection.to));
						}), e.focus();
					},
					children: s.comments.save_button_text
				})
			})
		})
	});
}
//#endregion
//#region src/components/Comments/FloatingComposerController.tsx
var b = /* @__PURE__ */ e({ default: () => x });
function x(e) {
	let r = i(), a = t(f), c = s(f, {
		editor: r,
		selector: (e) => e.pendingComment
	}), l = o({
		editor: r,
		selector: ({ editor: e }) => c ? {
			from: e.prosemirrorState.selection.from,
			to: e.prosemirrorState.selection.to
		} : void 0
	}), u = d(() => ({
		...e.floatingUIOptions,
		useFloatingOptions: {
			open: !!c,
			onOpenChange: (e) => {
				e || (a.stopPendingComment(), r.focus());
			},
			placement: "bottom",
			middleware: [
				m(10),
				h(),
				p()
			],
			...e.floatingUIOptions?.useFloatingOptions
		},
		focusManagerProps: { disabled: !1 },
		elementProps: {
			style: { zIndex: 60 },
			...e.floatingUIOptions?.elementProps
		}
	}), [
		a,
		r,
		c,
		e.floatingUIOptions
	]), g = e.floatingComposer || y;
	return /* @__PURE__ */ _(n, {
		position: l,
		portalElement: e.portalElement,
		...u,
		children: /* @__PURE__ */ _(g, {})
	});
}
//#endregion
export { b as n, y as r, x as t };
